### ISSUE TITLE
is [["sigmaTheta"]] obeying oceEOS setting?

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#1256]
Some code is given below (and graph attached). I would have expected the curves to either match perfectly or to mainly match at the surface.  After all, this is surface-referenced, not deep-referenced. I am wondering whether maybe the ``[[`` method is not checking the `oceEOS` setting.  I'm likely look at this within the next few days.
```R
library(oce)
data(section)
s100 <- section[["station", 100]]
options(oceEOS="gsw")
A <- s100[["sigmaTheta"]]
options(oceEOS="unesco")
B <- s100[["sigmaTheta"]]
A-B
```
yields
```R
 [1] 0.0039338419 0.0042496729 0.0045493035 0.0045721132 0.0045547276 0.0039969733 0.0026693163 0.0012177162 0.0008719962
[10] 0.0008460570 0.0008230778 0.0008102765 0.0008023929 0.0007871414 0.0007768708 0.0007756464 0.0007782389 0.0007857467
[19] 0.0008003401 0.0008210164 0.0008394563 0.0008589066 0.0008685607 0.0008621416
```
and
```R
plot(A-B,s100[["pressure"]],ylim=rev(range(s100[["pressure"]])))
```
yields as shown below.
![screen shot 2017-05-31 at 12 37 36 pm](https://cloud.githubusercontent.com/assets/99469/26640383/f4dd9d0e-45fd-11e7-90c5-291e113ad0cf.png)

**Comment by dankelley** [comment]
I started instrumenting the code and am suspecting that my logic on the gsw case is simply wrong.  What I'm doing (in `R/sw.R` near line 1536) is
```R
        theta <- swTheta(salinity=l$salinity, temperature=l$temperature, pressure=l$pressure,
                         referencePressure=referencePressure,
                         longitude=l$longitude, latitude=l$latitude, eos=l$eos)
        swRho(salinity=l$salinity, temperature=theta, pressure=referencePressure,
              longitude=l$longitude, latitude=l$latitude, eos=l$eos) - 1000
 ```
In other words, I'm relying on `swTheta` to be working right, and for the logic of the "old way" (compute theta, use that for temperature) will work in gsw. But in gsw, we have a new way, and I no longer see any point in fiddling with the old way. The new way is
```R
SA <- ...
CT <- ...
gsw_rho(SA, CT, 0)
```

**Comment by dankelley** [comment]
I've pushed to `develop` and to the test suite, so now `1256` in the later does some testing. I see differences even with this "new" way I just mentioned. I feel that maybe CR and I should discuss this in person sometime. For one thing, the gsw way of getting a potential density is not really too clearly explained in their webpages (in contrast to the unesco system, where things seemed clear). I also don't know just how different to expect the two sigma-theta values to be (is 0.01 percent, which is what I see, something to be expected, given that the old theta calculation was just one of several published algorithms?)

I would like to find some definitive statement about how gsw suggests that sigmaTheta be calculated, ideally with test values that can be burned into the test suite.

So, I am proposing to leave this until CR and DK have a f2f.

**Comment by dankelley** [comment]
Oh, hang on. I just skimmed through the TEOS-10_Manual.pdf and see on p197 that we have `gsw_pot_rho_t_exact` also.  That would seem to be the simplest approach to take. Unfortunately ... it gives a third answer, different to the unesco (as expected) and different to `gsw_rho()` (not as expected ...).

**Comment by richardsc** [comment]
I'm also confused a bit about this ... For one thing, `sigmaTheta` doesn't really exist in TEOS because the new temperature is CT. But does that mean that we wouldn't expect `swSigmaTheta()` and `gsw_rho()` to give the same thing? `gsw_rho()` is referred to as "potential density", e.g.

http://teos-10.org/pubs/gsw/html/gsw_rho.html

which requires `SA` as an input (and is therefore dependent on longitude and latitude). Example:
```r
> SA <- gsw_SA_from_SP(35, 0, 0, 0)
> CT <- gsw_CT_from_t(SA, 15, 0)
> gsw_rho(SA, CT, 0)-1000
[1] 25.97671
> swSigmaTheta(35, 15, 0)
[1] 25.97117
> (gsw_rho(SA, CT, 0)-1000 - swSigmaTheta(35, 15, 0))/swSigmaTheta(35, 15, 0)*100
[1] 0.02131112
```
So, a difference at the 0.02% level (for lon=lat=0). Not sure how that actually compares to measurement accuracy ...

**Comment by richardsc** [comment]
As for `gsw_pot_rho_t_exact()`, you mean this:

http://teos-10.org/pubs/gsw/html/gsw_pot_rho_t_exact.html

And yes, it gives a different value from `gsw_rho()` ...
```r
> gsw_pot_rho_t_exact(SA, 15, 0, 0)
[1] 1025.976
> gsw_rho(SA, CT, 0)
[1] 1025.977
> gsw_rho(SA, CT, 0) == gsw_pot_rho_t_exact(SA, 15, 0, 0)
[1] FALSE
> (gsw_rho(SA, CT, 0) - gsw_pot_rho_t_exact(SA, 15, 0, 0))/gsw_pot_rho_t_exact(SA, 15, 0, 0)
[1] 2.279466e-07
> (gsw_rho(SA, CT, 0) - gsw_pot_rho_t_exact(SA, 15, 0, 0))/gsw_pot_rho_t_exact(SA, 15, 0, 0)*100
[1] 2.279466e-05
```

**Comment by dankelley** [comment]
I've pushed to `develop` and the test suite (I know you know where it is, so I won't open a browser so I can find the link to put here) and it now also spits out graphs.  I think the teos-10 docs are confusing on this; the table on p197 would make a person think that there were 2 different ways of doing this.

One thing we *do* know: `gsw_pot_rho_t_exact` is clearly documented, and we recover its test values (as tested in building `gsw`). So I feel that's the way to go.  If you rebuild and run the test in `1256` you'll get an Rplot.pdf that summarizes.  It seems that the gsw is diff from the unesco by about 0.005 kg/m^3 in this particular example.  Now, the thing is -- the unesco approach is kind of wonky, using a certain (Bryden) formula for theta, even though there are alternatives. (This has always bothered me, that the theta work and literature seemed independent of the rho literature.)

I think we might want to talk about this in person sometime. Not much rush, though. And I think we want to be quite clear on the contradictions between the two gsw approaches, before we make an issue for the teos-10 people ... although I do think we'll want to make such an issue, unless we find documentation that  these two approaches are actually meant to differ. (I wonder whether some of this is the 'exact' versus 75-term, for example. I have no idea what 'exact' really means in gsw parlance)

**Comment by dankelley** [comment]
**Status report.** I flipped through the TEOS-10 documents and did not see a lot about the different ways of computing density. The text is not entirely tied to the functions of the package. That means we are left to decided whether to compute sigma0 (say) by using `gsw_pot_rho_t_exact(SA,t,p,p_ref)` or `gsw_rho(SA,CT,p_ref)`. The entirety of the docs for the former [1] suggest that it's the right method, while the docs for the latter [2] suggest that it can be used for this

> Note that potential density with respect to reference pressure, p_ref, is
obtained by calling this function with the pressure argument being p_ref
(i.e. "gsw_rho(SA,CT,p_ref)").

but, as @richardsc and I have both found (in R, anyway) the functions do *not* give the same result. In the oce context, I am not terribly concerned about this seeming contradiction in GSW.  We know that our test values (from the GSW matlab docs) work, so we know we are giving the user access to the GSW functions correctly. 

PROPOSAL 1: Use `gsw_pot_rho_t_exact` and to document it that way. (I think after a f2f between me and CR, we might want to submit an issue to teos-10, but that's another matter.)

PROPOSAL 2: disallow `swTheta` and `swSigmaTheta` with `eos="gsw"`, reporting an error that explains what they ought to do instead. This one is more controversial, and I'd particularly like to know what @richardsc thinks.  My reasoning is that gsw does not provide these things, and providing some sort of half-equivalent will just confuse matters. If the user wants gsw, they just have to live with that choice and give up on theta and sigmaTheta, etc.  The downside is that this may break code ... it will not be possible for a user to put `options(oceEOS="gsw")` at the top of a script and expect everything to work properly ... well, it will be proper, so long as there are no direct calls to `swTheta` etc. This is something that @richardsc and I might want to discuss in person, unless he is in full agreement with my proposal. (If he *is* in agreement, I can do this on the weekend.)

1. http://www.teos-10.org/pubs/gsw/html/gsw_pot_rho_t_exact.html
2. http://teos-10.org/pubs/gsw/html/gsw_rho.html

**Comment by dankelley** [comment]
I've done some reading and thinking, and I think I've made the right decisions in recoding some `swTheta` and `swSigmaTheta` actions. I have followed up on my proposal in the previous comment. Now, the issue 1256 test code makes a graph like shown below. I think this 0.01 percent difference between the two formulations may just be the fact that the formulations are a bit different. In any case, I'm going to close this issue now, since I think it's OK. If @richardsc wants to reopen it or discuss it when we meet next, that's fine, of course.

![rplots](https://cloud.githubusercontent.com/assets/99469/26756935/aca94c76-4883-11e7-853b-419f74487c28.png)

**Comment by richardsc** [comment]
It's going to be awhile before we can have a face-to-face, as I'm headed to CMOS later today. I was going to agree with both proposal 1 and 2 -- i.e. remove the ability for `swTheta()` and `swSigmaTheta()` to have an `eos` option, since as defined they really only have meaning under the UNESCO formulation. Is that what you're referring to? I notice from a pull and rebuild this morning that:
```r
swSigmaTheta(35, 15, 0, eos='gsw', longitude=0, latitude=0)
```
produces
```
[1] 25.97647
```
so maybe I'm confused about proposal 2.

**Comment by dankelley** [comment]
No worries on an f2f. Enjoy CMOS!!

Actually, it turns out that gsw does have a "theta", but without that name. It's called `pt` as in http://www.teos-10.org/pubs/gsw/html/gsw_pt_from_t.html and I just hadn't noticed it before. So, I did a U-turn in my coding yesterday: first I disallowed the `eos` arg for `swTheta` and `swSigmaTheta`, and then I undid that code and allowed it again. I didn't put enough comments here to explain my thought process, partly because I had not pushed the "disallow" code to GH.

I can make it clearer by changing the first line of the `swTheta` doc from
> Compute theta, the potential temperature of seawater.
to
> Compute the potential temperature of seawater.

and I think I should, because in GSW, there *is* nothing called "theta".  (I don't know what symbol they recommend for this quantity.)

NOTE: I don't really think `CT` is a replacement for theta, because it is not depth-referenced. (I admit, though, that the majority of dynamical calculations involve density, not any sort of temperature, so theta isn't that useful except for display purposes. The exception, where temperature actually matters, would be to biology, chemistry, and things like double-diffusion ... but in those cases in-situ is likely the important thing, unless we imagine some scenario that moves water kilometres in the vertical.)
